src/diceBot/DoubleCross.rb

```diff
@@ -37,32 +34,182 @@ def getHelpMessage
+      return targetValue ? "#{lhs}>=#{targetValue}" : lhs
+    end
+
+    # 出力用に整形された修正値を返す
+    # @return [String]
+    def formattedModifier
+      if modifier == 0
+        ''
+      elsif modifier > 0
+        "+#{modifier}"
+      else
+        modifier.to_s
+      end
+    end
+  end
+
+  # 出目のグループを表すクラス
+  class ValueGroup
+    # 出目の配列
+    # @return [Array<Integer>]
+    attr_reader :values
+    # クリティカル値
+    # @return [Integer]
+    attr_reader :criticalValue
+
+    # 出目のグループを初期化する
+    # @param [Array<Integer>] values 出目の配列
+    # @param [Integer] criticalValue クリティカル値
+    def initialize(values, criticalValue)
+      @values = values.sort
+      @criticalValue = criticalValue
+    end
 
-    debug("DoubleCross parren_killer_add string", string)
+    # 出目のグループの文字列表記を返す
+    # @return [String]
+    def to_s
+      "#{max}[#{@values.join(',')}]"
+    end
 
-    string = string.gsub(/(\d+)DX(\d*)([^\d\s][\+\-\d]+)/i) { "#{Regexp.last_match(1)}R10#{Regexp.last_match(3)}[#{Regexp.last_match(2)}]" }
-    string = string.gsub(/(\d+)DX(\d+)/i) { "#{Regexp.last_match(1)}R10[#{Regexp.last_match(2)}]" }
-    string = string.gsub(/(\d+)DX([^\d\s][\+\-\d]+)/i) { "#{Regexp.last_match(1)}R10#{Regexp.last_match(2)}" }
-    string = string.gsub(/(\d+)DX/i) { "#{Regexp.last_match(1)}R10" }
-    if /\@(\d+)/ =~ string
-      crit = Regexp.last_match(1)
-      string = string.gsub(/\[\]/) { "\[#{crit}\]" }
-      string = string.gsub(/\@(\d+)/, "")
+    # 出目のグループ中の最大値を返す
+    # @return [Integer]
+    #
+    # クリティカル値以上の出目が含まれていた場合は10を返す
+    def max
+      @values.any? { |value| critical?(value) } ? 10 : @values.max
     end
-    string = string.gsub(/\[\]/, "")
 
-    debug("DoubleCross parren_killer_add changed string", string)
+    # クリティカルの発生数を返す
+    # @return [Integer]
+    def numOfCriticalOccurrences
+      @values.
+        select { |value| critical?(value) }.
+        length
+    end
 
-    return string
+    # クリティカルが発生したかを返す
+    # @param [Integer] value 出目
+    # @return [Boolean]
+    def critical?(value)
+      value >= @criticalValue
+    end
   end
 
-  def dice_command_xRn(string, nick_e)
-    output_msg = check_dice(string)
-    return nil if output_msg.nil?
+  # 成功判定コマンドのダイスロール結果を表すクラス
+  class DXDiceRollResult
+    # 成功判定コマンドのノード
+    # @return [DXNode]
+    attr_reader :dxNode
+    # 出目のグループの配列
+    # @return [Array<ValueGroup>]
+    attr_reader :valueGroups
+    # 回転数
+    # @return [Integer]
+    attr_reader :loopCount
+
+    # 達成値
+    # @return [Integer]
+    attr_reader :achievedValue
+    # ファンブルかどうか
+    # @return [Boolean]
+    attr_reader :isFumble
+
+    # 成功判定コマンドのダイスロール結果を初期化する
+    # @param [DXNode] dxNode 成功判定コマンドのノード
+    # @param [Array<ValueGroup>] valueGroups 出目のグループの配列
+    # @param [Integer] loopCount 回転数
+    def initialize(dxNode, valueGroups, loopCount)
+      @dxNode = dxNode
+      @valueGroups = valueGroups
+      @loopCount = loopCount
+
+      sum = @valueGroups.reduce(0) { |acc, group| acc + group.max }
```

Comment:
```suggestion
      sum = @valueGroups.map(&:max).reduce(:+)
```
この様にも書けますがどちらが見やすいかは謎です。

Reply:
v3で最近のRuby（2.4以降）になれば `Array#sum` が使えるようになってより分かりやすくなるので、`map` を使う方にしたいと思います。

`reduce` を使う際には、初期値を伴う `.reduce(0, :+)` の形にしておきたいと思います。これは、空の配列に対して呼ぶと `nil` が返り、後の計算で例外が発生し得るためです。ダイス数0を弾いているのであり得ないとは思いますが、型が合うように注意しておきたいです。

Reply:
3deb9b5 で変更しました。